New module `build_schema`

Daniel-VM opened the pull request:
## PR Description

This pull request introduces a new module, build_schema.py, designed to parse a database definition in XLSX format and convert it into two files simultaneously: relecov schema definition (relecov_tools/schema/relecov_schema.json) and Metadata-Lab Template (XLSX file template for entering lab metadata).

Additionally, this module will generate and save a report that compares the updated version of relecov_schema.json with its previous version after each execution of the module, while also tracking versioning.

Closes #259

Shettland commented on relecov_tools/build_schema.py:
```diff
@@ -0,0 +1,206 @@
+
+        # Convert database to json format
+        json_data = {}
+        for _, row in df.iterrows():
```
I think that for this case you can use itertuples() which is [significantly faster](https://stackoverflow.com/questions/70227908/iterating-over-rows-in-a-dataframe-in-pandas-is-there-a-difference-between-usin)

Shettland commented on relecov_tools/__main__.py:
```diff
@@ -473,5 +474,34 @@ def metadata_homogeneizer(institution, directory, output):
+    "-s",
+    "--schema_base",
+    type=click.Path(),
+    help="Path to the Excel document containing the database definition.",
```
This description is duplicated

Shettland commented on relecov_tools/build_schema.py:
```diff
@@ -0,0 +1,589 @@
+        """Validate the mandatory features of each property in json_data."""
+        # Check mandatory key features to build a json schema
+        notvalid_properties = {}
+        mandatory_features = [
+            "enum",
+            "examples",
+            "ontology_id",
+            "type",
+            "description",
+            "classification",
+            "label_name",
+            "fill_mode",
+            "required (Y/N)",
+            "complex_field (Y/N)",
+        ]
```
These mandatory fields should be saved somewhere in the user guide or readme

Shettland reviewed (changes requested):
Most functions are missing the extended docstrings, including descriptions for args and returns. I'd say that they are needed at least for the ones in `build_schema.py`.

Shettland commented on relecov_tools/build_schema.py:
```diff
@@ -0,0 +1,589 @@
+        return draft_template
+
+    def standard_jsonschema_object(self, data_dict, target_key):
+        """ "Create standar json schema object"""
```
Extra `"` here

Shettland commented on relecov_tools/assets/schema_utils/metadatalab_template.py:
```diff
@@ -0,0 +1,145 @@
+
+        # setup excel format
+        worksheet.set_column(0, len(df.columns), 30)
+        header_formater = workbook.add_format(
```
A nice addition could be to optionally lock the header and first column so they cannot be modified by the users. Reference on how to set a custom password with protect() and add_format() here: https://xlsxwriter.readthedocs.io/format.html#set_locked

Shettland replied:
Another good feature would be to already include data validation here, giving the users only a certain list of fields to choose from for each columns, based on the enums of the corresponding property from the processed schema: https://xlsxwriter.readthedocs.io/working_with_data_validation.html#working-with-data-validation

Daniel-VM replied:
I attempted to add the key 'locked': True as specified in the documentation, but it didn’t work. This issue may require further investigation.
```
header_formater = workbook.add_format(
            {
                "bold": True,
                "text_wrap": False,
                "valign": "top",
                "fg_color": "#ADD8E6",
                "border": 1,
                "locked": True,
            }
        )
first_col_formater = workbook.add_format(
            {
                "bold": True,
                "text_wrap": False,
                "valign": "center",
                "fg_color": "#ADD8E6",
                "border": 1,
                "locked": True,
            }
        )
```

Shettland commented on requirements.txt:
```diff
@@ -10,6 +10,8 @@ pyyaml
-tabulate
\ No newline at end of file
+tabulate
+cdifflib==1.2.6
```
I only found python's built-in `difflib`, were you planning to use that?

Daniel-VM replied:
I forgot to remove `cdifflib` from the requirements. I ended up using the built-in difflib instead.

saramonzon commented:
Some minor stuff:
- When you use `-o` param it stills ask you where you want to output, and creates two levels of subfolders `results/results/files..`
- The param `-d true` should be use just with `-d` without the true, that is a param in the args functionality

Something I do not understand:
what does this mean in the standard output? It's not very informative
```
Schema validation error: '' is not valid under any of the given schemas
```

Another thing to take into account there is no log outputted when using the `--log-file` core functionality, this can be a follow up PR but make and issue because it's important to save log information about the build schema process.

but it looks really nice @Daniel-VM !! And the testing data you provided really helpful for testing and checking everything out, thank you so much!

Daniel-VM commented:
- [X] Added docstrings to  build_schema.py
- [X] Fixed output folder duplication
- [X] Added -d as flag param
- [X] Recover raw json validation stdout to make more trazable the validation step
- [X] Fixed code overlapping when validating schemas
- [ ] Created an Issue to implement --log-file core functionality.

@saramonzon @Shettland

saramonzon commented on relecov_tools/__main__.py:
```diff
@@ -21,7 +21,11 @@
 import relecov_tools.metadata_homogeneizer
 import relecov_tools.gisaid_upload
 import relecov_tools.upload_ena_protocol
+<<<<<<< HEAD
```
Conflict wrongly commited!

Shettland reviewed (approved):
Amazing job, looks really good!
